pkg/coscheduling/coscheduling.go

```diff
@@ -95,295 +84,374 @@ func New(obj runtime.Object, handle framework.FrameworkHandle) (framework.Plugin
+		if err == util.ErrorNotMatched {
+			return framework.NewStatus(framework.Success, ""), 0
+		}
+		klog.Infof("cs.pgMgr.Permit error %v", err)
+		return framework.NewStatus(framework.Unschedulable, err.Error()), waitTime
+	}
+	var pgStatus *schecache.PodGroupSchedulingStatus
+	fullName := util.GetPodGroupFullName(pod)
+	if pgStatus = cs.pgMgr.PodGroupStatusCache().Get(fullName); pgStatus == nil {
+		return framework.NewStatus(framework.UnschedulableAndUnresolvable, "PodGroup not found"), waitTime
+	}
+	if wait := util.GetWaitTimeDuration(pgStatus.PodGroup, cs.scheduleTimeout); wait != 0 {
+		waitTime = wait
+	}
+	klog.V(5).Infof("Pod requires pgName %v", fullName)
+	if ready {
+		cs.sendStartScheduleSignal(fullName)
+	}
+	return framework.NewStatus(framework.Wait, ""), waitTime
 }
 
-func (cs *Coscheduling) Reserve(ctx context.Context, state *framework.CycleState, p *v1.Pod, nodeName string) *framework.Status {
+func (cs *Coscheduling) Reserve(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) *framework.Status {
 	return nil
 }
 
 // Unreserve rejects all other Pods in the PodGroup when one of the pods in the group times out.
 func (cs *Coscheduling) Unreserve(ctx context.Context, state *framework.CycleState, pod *v1.Pod, nodeName string) {
-	pgInfo, _ := cs.getOrCreatePodGroupInfo(pod, time.Now())
-	if len(pgInfo.key) == 0 {
+	_, pgStatus := cs.pgMgr.GetPodGroup(pod)
+	if pgStatus == nil {
 		return
 	}
-	podGroupName := pgInfo.name
-	cs.frameworkHandle.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
-		if waitingPod.GetPod().Namespace == pod.Namespace && waitingPod.GetPod().Labels[PodGroupName] == podGroupName {
-			klog.V(3).Infof("Unreserve rejects the pod: %v/%v", podGroupName, waitingPod.GetPod().Name)
-			waitingPod.Reject(cs.Name())
+	pgStatus.AssumedPods.Delete(core.GetNamedspacedName(pod))
+}
+
+// PostBind is called after a pod is successfully bound. These plugins are used update PodGroup when pod is bound.
+func (cs *Coscheduling) PostBind(ctx context.Context, _ *framework.CycleState, pod *v1.Pod, nodeName string) {
+	klog.V(5).Infof("PostBind pod: %v", core.GetNamedspacedName(pod))
+	cs.pgMgr.PostBind(ctx, pod, nodeName)
+}
+
+// VerifyLocalCache verify pgStatustatus cache
+func (cs *Coscheduling) VerifyLocalCache(toSchedulePGName string) bool {
+	pgStatus := cs.pgMgr.PodGroupStatusCache().Get(toSchedulePGName)
+	podInCacheUpdate := false
+	// verify pods cache
+	cs.frameworkHandler.IterateOverWaitingPods(func(waitingPod framework.WaitingPod) {
+		klog.V(5).Info("Start IterateOverWaitingPods")
+		pod := waitingPod.GetPod()
+		pgName := core.GetNamedspacedName(pgStatus.PodGroup)
+		klog.V(5).Infof("Start walking through pod %v uid: %v", core.GetNamedspacedName(pod), pod.UID)
+		if pgName == toSchedulePGName {
+			klog.V(5).Infof("Wanted cache pod %v", core.GetNamedspacedName(pod))
+			_, found := pgStatus.AssumedPods.Get(core.GetNamedspacedName(pod))
```

Comment:
Does this mean some pod belongs to the PodGroup gets removed from the cache? If so, can you comment above when this would happen?

Reply:
This may happen when pod delete and re-create. 

Comments added